src/main/java/com/google/cloud/dataproc/templates/config/GeneralTemplateConfig.java

```diff
@@ -0,0 +1,157 @@
+import jakarta.validation.Constraint;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import jakarta.validation.ConstraintViolation;
+import jakarta.validation.Payload;
+import jakarta.validation.Valid;
+import jakarta.validation.Validation;
+import jakarta.validation.Validator;
+import jakarta.validation.ValidatorFactory;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.NotNull;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@GeneralTemplateConfigAnnotation
+public class GeneralTemplateConfig {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(GeneralTemplateConfig.class);
+
+  @NotEmpty() private Map<String, @Valid InputConfig> input = ImmutableMap.of();
+  @NotNull private Map<String, @Valid QueryConfig> query = ImmutableMap.of();;
+  @NotEmpty() private Map<String, @Valid OutputConfig> output = ImmutableMap.of();;
+
+  public Map<String, InputConfig> getInput() {
+    return input;
+  }
+
+  public void setInput(Map<String, InputConfig> input) {
+    this.input = input;
+  }
+
+  public Map<String, OutputConfig> getOutput() {
+    return output;
+  }
+
+  public void setOutput(Map<String, OutputConfig> output) {
+    this.output = output;
+  }
+
+  public Map<String, QueryConfig> getQuery() {
+    return query;
+  }
+
+  public void setQuery(Map<String, QueryConfig> query) {
+    this.query = query;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("input", input)
+        .add("output", output)
+        .add("query", query)
+        .toString();
+  }
+
+  public Set<ConstraintViolation<GeneralTemplateConfig>> validate() {
+    return validate(this);
+  }
+
+  static <T> Set<ConstraintViolation<T>> validate(T pojo) {
+    ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
+    Validator validator = factory.getValidator();
+    Set<ConstraintViolation<T>> violations = validator.validate(pojo);
+    for (ConstraintViolation<T> violation : violations) {
+      LOGGER.error(
+          "Violation: {}; {}; in {}",
+          violation.getPropertyPath(),
+          violation.getMessage(),
+          violation.getRootBeanClass());
+    }
+
+    return violations;
+  }
+
+  @Constraint(validatedBy = GeneralTemplateConfigValidator.class)
+  @Target(ElementType.TYPE)
+  @Retention(RetentionPolicy.RUNTIME)
+  public @interface GeneralTemplateConfigAnnotation {
+
+    String message() default "inconsistent config";
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
+  }
+
+  public static class GeneralTemplateConfigValidator
```

Comment:
This is custom annotation validator, the `@NonEmpty` `@NotNull` constraints are checked regardless (and in the child pojos, this validator checks that the keys in the input, query and output maps don't conflict or match as required.